command_test.go

```diff
@@ -800,4 +823,57 @@ func TestCommand_DefaultsOverride(t *testing.T) {
 
 		inv.Args = []string{"--config", fi.Name()}
 	})
+
+	test("AltFlagOverDefault", "good.com", func(t *testing.T, inv *serpent.Invocation) {
+		inv.Args = []string{"--url-deprecated", "good.com"}
+	})
+}
+
+func TestCommand_OptionsWithSharedValue(t *testing.T) {
+	t.Parallel()
+
+	var got string
+	makeCmd := func(def, altDef string) *serpent.Command {
+		got = ""
+		return &serpent.Command{
+			Options: serpent.OptionSet{
+				{
+					Name:    "url",
+					Flag:    "url",
+					Default: def,
+					Value:   serpent.StringOf(&got),
+				},
+				{
+					Name:    "alt-url",
+					Flag:    "alt-url",
+					Default: altDef,
+					Value:   serpent.StringOf(&got),
+				},
+			},
+			Handler: (func(i *serpent.Invocation) error {
+				return nil
+			}),
+		}
+	}
+
+	// Check proper value propagation.
+	err := makeCmd("def.com", "def.com").Invoke().Run()
+	require.NoError(t, err, "default values are same")
+	require.Equal(t, "def.com", got)
+
+	err = makeCmd("def.com", "").Invoke().Run()
+	require.NoError(t, err, "other default value is empty")
+	require.Equal(t, "def.com", got)
+
+	err = makeCmd("def.com", "").Invoke("--url", "sup").Run()
+	require.NoError(t, err)
+	require.Equal(t, "sup", got)
+
+	err = makeCmd("def.com", "").Invoke("--alt-url", "hup").Run()
+	require.NoError(t, err)
+	require.Equal(t, "hup", got)
```

Comment:
Should we try to handle cases where both are set? Currently it defaults to the last one set. Should we at least make this deterministic, like by alphabetical order? 

I have not tested if 1 is set with an env var, the other the flag.

```
// Passes
err = makeCmd("def.com", "").Invoke("--url", "sup", "--alt-url", "hup").Run()
require.NoError(t, err)
require.Equal(t, serpent.String("hup"), got)

// Fails
err = makeCmd("def.com", "").Invoke("--alt-url", "hup", "--url", "sup").Run()
require.NoError(t, err)
require.Equal(t, serpent.String("hup"), got)
```

Reply:
I don't think alphabetical order is going to be intuitive. The current behaviour of "last one wins" makes more sense to me.

Alternatively, we could say that specifying a non-default value for both is an error? That would remove any ambiguity, but would be a breaking change.

Reply:
@Emyrk it's typical behavior for CLI programs to use the last value of any given flag, for example:

```
~/Work/Code/serpent mafredri/fix-shared-value-overwrite*
❯ git log -1 --format='%h'
e053fda

~/Work/Code/serpent mafredri/fix-shared-value-overwrite*
❯ git log -1 --format='%h' --format='%H'
e053fdac0b0dd01a0e6278dfbd703bbf7486a840
```

So I would say the failing test-case is working as intended, just the assertion that should change.

I added some more test-cases to cover this and also the env/flag priority one in 457cbbe.

@johnstcn In this PR we already handle the case where flags specify multiple defaults, and that's OK as long as:

- Either only one specifies a default
- Either one or more specify the same exact default
- If two different defaults are detected, that's an error

Reply:
This made me curious. In the case of envs, it seems to be "first wins". I tested with the following test-cases.

```go
// Both envs are given, first wins.
inv = makeCmd("def.com", "").Invoke()
inv.Environ.Set("URL", "sup")
inv.Environ.Set("ALT_URL", "hup")
err = inv.Run()
require.NoError(t, err)
require.Equal(t, "sup", got)

// Both envs are given, first wins #2.
inv = makeCmd("def.com", "").Invoke()
inv.Environ.Set("ALT_URL", "hup")
inv.Environ.Set("URL", "sup")
err = inv.Run()
require.NoError(t, err)
require.Equal(t, "sup", got)
```

I'm not sure if we need to do anything about that, but just keep in mind that this behavior has not changed in this PR.